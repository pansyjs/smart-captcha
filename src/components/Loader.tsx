import { useExternal } from '@pansy/use-external';
import React from 'react';

export interface LoaderProps {
  children?: React.ReactNode;
  loading?: React.ReactNode;
}

export function Loader(props: LoaderProps) {
  const { children, loading } = props;
  const status = useExternal('//g.alicdn.com/AWSC/AWSC/awsc.js');

  if (status === 'loading' && loading) {
    return loading;
  }

  return (
    <>
      {(status === 'ready') && children}
    </>
  );
}
