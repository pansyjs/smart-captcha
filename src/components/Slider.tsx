import React, {
  forwardRef,
  useRef,
  useEffect,
  useImperativeHandle,
} from 'react';

import type { SmartCaptchaResult, SmartCaptchaRef, SliderCaptchaProps } from '../types';

const SliderCaptcha: React.ForwardRefRenderFunction<SmartCaptchaRef, SliderCaptchaProps> = (
  {
    className,
    style,
    elementId = 'slider-captcha',
    appkey,
    scene,
    foreign,
    width,
    language,
    upLang,
    hideErrorCode,
    fontSize,
    onSuccess,
    onChange,
    onFailed,
    onError,
  },
  ref,
) => {
  const ic = useRef<any>();

  useImperativeHandle(ref, () => ({
    reset: () => {
      ic?.current?.reset();
    },
  }));

  useEffect(() => {
    if (window.AWSC && !ic.current) {
      window.AWSC.use('nc', (_: string, module: any) => {
        ic.current = module.init({
          appkey,
          scene,
          renderTo: `${elementId}`,
          // 智能验证组件的宽度。
          width: width,
          foreign,
          language,
          upLang,
          fontSize,
          hideErrorCode,
          success: function(data: SmartCaptchaResult) {
            const value = {
              sig: data.sig,
              token: data.token,
              sessionId: data.sessionId,
              appKey: appkey,
              scene,
            }
            onChange?.(value);
            onSuccess?.(value);
          },
          fail: function() {
            ic?.current?.reset();
            onChange?.(undefined);
            onFailed?.();
          },
          error: function() {
            ic?.current?.reset();
            onChange?.(undefined);
            onError?.();
          },
        })
      })
    }
  }, []);

  return (
    <div className={className} style={style} id={elementId} />
  );
};

export default forwardRef(SliderCaptcha);
