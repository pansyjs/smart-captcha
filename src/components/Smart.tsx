import React, {
  forwardRef,
  useRef,
  useEffect,
  useImperativeHandle,
} from 'react';
import { Option } from '../config';

import type { SmartCaptchaProps, SmartCaptchaResult, SmartCaptchaRef, } from '../types';

const SmartCaptcha: React.ForwardRefRenderFunction<SmartCaptchaRef, SmartCaptchaProps> = (
  {
    className,
    style,
    elementId = 'smart-captcha',
    appkey = Option.appkey,
    scene = Option.scene,
    width = 300,
    height = 42,
    defaultTxt = '点击按钮开始智能验证',
    successTxt = '验证成功',
    scaningTxt = '智能检测中',
    failTxt = '验证失败，请在此点击按钮刷新',
    language,
    upLang,
    onSuccess,
    onChange,
    onFailed,
  },
  ref,
) => {
  const ic = useRef<any>();

  useImperativeHandle(ref, () => ({
    reset: () => {
      ic?.current?.reset();
    },
  }));

  useEffect(() => {
    if (window.AWSC) {
      window.AWSC.use('ic', (_: string, module: any) => {
        if (ic.current) return;
        ic.current = module.init({
          // 声明智能验证需要渲染的目标元素ID。
          renderTo: `#${elementId}`,
          // 智能验证组件的宽度。
          width: width,
          // 智能验证组件的高度。
          height: height,
          // 智能验证组件初始状态文案。
          default_txt: defaultTxt,
          // 智能验证组件验证通过状态文案。
          success_txt: successTxt,
          // 智能验证组件验证失败（拦截）状态文案。
          fail_txt: failTxt,
          // 智能验证组件验证中状态文案。
          scaning_txt: scaningTxt,
          // 应用类型标识
          appkey,
          // 使用场景标识
          scene,
          language,
          upLang,
          success: function(data: SmartCaptchaResult) {
            const value = {
              sig: data.sig,
              token: data.token,
              sessionId: data.sessionId,
              appKey: Option.appkey,
              scene: Option.scene,
            }
            onChange?.(value);
            onSuccess?.(value);
          },
          fail: function() {
            ic?.current?.reset();
            onFailed?.();
            onChange?.(undefined);
          },
          error: function() {
            ic?.current?.reset();
            onFailed?.();
            onChange?.(undefined);
          },
        })
      })
    }

    return () => {
      ic.current = undefined;
    }
  }, []);

  return (
    <div className={className} style={style} id={elementId} />
  );
};

export default forwardRef(SmartCaptcha);
