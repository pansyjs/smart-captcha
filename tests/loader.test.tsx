import { act, fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Loader } from '../src/components/Loader';

describe('apiLoader', () => {
  it('should be defined', () => {
    expect(Loader).toBeDefined();
  });

  it('renders correctly', async () => {
    const { container } = render(<Loader>children</Loader>);

    const script = document.querySelector('script')!;

    await act(() => {
      fireEvent.load(script);
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
