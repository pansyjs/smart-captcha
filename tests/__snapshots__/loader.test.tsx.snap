// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiLoader renders correctly 1`] = `children`;
